cgo/lapack.go

```diff
@@ -108,6 +111,65 @@ type Implementation struct{}
 
 var _ lapack.Float64 = Implementation{}
 
+// Dgeqp3 computes a QR factorization with column pivoting of the
+// m×n matrix A: A*P = Q*R using Level 3 BLAS.
+//
+// The matrix Q is represented as a product of elementary reflectors
+//  Q = H_0 H_1 . . . H_k-1, where k = min(m,n).
+// Each H[i] has the form
+//  H[i] = I - tau * v * v^T
+// where tau is a real vector, and v is a real/complex vector
+// with v[0:i-1] = 0 and v[i] = 1; v[i:m] is stored on exit in
+// A[i:m, i], and tau in tau[i].
+//
+// jpvt specifies a column pivot to be applied to A. If
+// jpvt[j] is at least zero, the jth column of A is permuted
+// to the front of A*P (a leading column), otherwise the jth
+// column of A is a free column. On return, jpvt holds the
+// permutation that was applied; if jpvt[j] == k, then the jth
+// column of A*P was the the kth column of A.
+// jpvt must have length n or Dgeqp3 will panic.
+//
+// On exit tau holds the scalar factors of the elementary reflectors.
+// It must have length n, otherwise Dgeqp3 will panic.
+//
+// work must have length max(1, lwork). otherwise Dgeqp3 will panic.
+// work must have length at least max(1,lwork), and lwork must be at least
+// 3*n+1. For optimal performance lwork must be at least 2*n+(n+1)*nb, where
+// nb is the optimal blocksize.On return, work[0] will contain the optimal
+// value of lwork.
+//
+// If lwork == -1, instead of performing Dgeqp3, only the optimal value of lwork
+// will be stored in work[0].
+//
+// Dgeqp3 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgeqp3(m, n int, a []float64, lda int, jpvt []int, tau, work []float64, lwork int) {
+	checkMatrix(m, n, a, lda)
```

Comment:
The rational I have for checking early is to fail if the parameters are wrong up front. I'm happy to change this, but I don't see a problem with this approach (I prefer it).